### PR TITLE
Add TrySingle<TSource>(IEnumerable<TSource>, out TSource) 

### PR DESCRIPTION
**Pull request by AlexRadch**
In LINQ we have the Single and SingleOrDefault methods, but they don't tell you if it succeeded or not.
See https://github.com/dotnet/corefx/issues/6073

### CODE REVIEW

**Comment by karelz** [comment]
@VSadov is OOF, @OmarTawfik @stephentoub can you please help with code review?

**Comment by AlexRadch** [comment]
I did not see compilation error for `netcoreapp1.1` tests and I did not see `netcoreapp1.1` failed tests if compile and run tests locally. How to compile and to run `netcoreapp1.1` tests locally?

When I compile all tests (`buld` in root folder), it compiles and run all test but it takes a long time.

When I compile tests locally `msbuild -t:Build` in `src\System.Linq\tests\` folder. It does not compile and does not run `netcoreapp1.1` tests.

File: src/System.Linq/src/System/Linq/Last.cs [hunk 1]
```diff
@@ -174,36 +103,48 @@ public static TSource LastOrDefault<TSource>(this IEnumerable<TSource> source, F
...
-                    if (predicate(element))
-                    {
-                        return element;
-                    }
+                    element = ordered.Last(predicate);
+                    return true;
                 }
-
-                return default(TSource);
+                catch (InvalidOperationException) { }
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Last.cs]: Same concern with the try/catch.  Should this be using TryGetLast?

**Comment by stephentoub** [comment]
cc: @JonHanna, @bartdesmet, @jamesqo

File: src/System.Linq/tests/ConsistencyTests.cs [hunk 2]
```diff
@@ -31,7 +31,11 @@ public static void MatchSequencePattern()
                      "LeftJoin",
                      "Append",
                      "Prepend",
-                     "ToHashSet"
+                     "ToHashSet",
+                     "TryElementAt",
```
Comment (JonHanna) [inline src/System.Linq/tests/ConsistencyTests.cs]: Why not implement on `Queryable`? (More a question for the issue than the PR).

File: src/System.Linq/src/System/Linq/Single.cs [hunk 3]
```diff
@@ -149,5 +150,79 @@ public static TSource SingleOrDefault<TSource>(this IEnumerable<TSource> source,
 
             return default(TSource);
         }
+
+        public static bool TrySingle<TSource>(this IEnumerable<TSource> source, out TSource element)
+        {
+            if (source == null)
```
Comment (AlexRadch) [inline src/System.Linq/src/System/Linq/Single.cs]: Single can throw `Error.NoElements()` or `Error.MoreThanOneElement()` exception. It is `InvalidOperationException` with different messages. 
When TrySingle return false, I did not know what exception `Error.NoElements()` or `Error.MoreThanOneElement()` should be.
I can de-duplicate this code only if add universal text message. I do not  know can I made such changes or can`t.
May be somebody later de-duplicate this code because I know english only a little.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Single.cs]: This is correct for the reason given, and part of the motivation of the issue that asked for this API change seemed to be precisely this difference in how multiple elements are handled.

**Comment by JonHanna** [comment]
@jamesqo's point about the name made recently at #6073 makes sense to me, and I'd also prefer to see Queryable covered at the same time. (Not able to examine the implementation of what's currently spec'd fully right now, but nothing stands out as a problem).

**Comment by AlexRadch** [comment]
@VSadov, @OmarTawfik I added implementation of Queryable.TrySingle only. Is my code correct? 
If it is correct I will make implementation of other `TryGet` methods.

File: src/System.Linq.Queryable/src/System/Linq/Queryable.cs [hunk 4]
```diff
@@ -842,6 +842,23 @@ public static TSource SingleOrDefault<TSource>(this IQueryable<TSource> source,
                     ));
         }
 
+        public static bool TrySingle<TSource>(this IQueryable<TSource> source, out TSource element)
+        {
+            var result = TrySingle(source);
```
Comment (karelz) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: Nit: Use var only when it is clear from the line what is the type -- see https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md [10]

Comment (AlexRadch) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: I'm trying to not use `var` but sometimes forget about it.

**Comment by bartdesmet** [comment]
At a quick glimpse, the Queryable implementation in the PR is very non-intuitive because it doesn't produce an expression referring to a publicly visible method, which is what query providers would look for. The EnumerableQuery/Rewriter support would also fail because there's no counterpart with the same signature on Enumerable. The use of a Tuple type seems arbitrary but once captured in an expression tree, it can't be changed without a breaking change (e.g. if we'd like to use ValueTuple instead, which would be cheaper, and arguably the better choice given its language-integrated status as the symmetric equivalent to input parameters). I'll give it some more thought how we should handle this.

File: src/System.Linq.Queryable/src/System/Linq/Queryable.cs [hunk 5]
```diff
@@ -842,6 +842,23 @@ public static TSource SingleOrDefault<TSource>(this IQueryable<TSource> source,
...
+        internal static Tuple<bool, TSource> TrySingle<TSource>(this IQueryable<TSource> source)
+        {
+            if (source == null)
+                throw Error.ArgumentNull(nameof(source));
+            return source.Provider.Execute<Tuple<bool, TSource>>(
+                Expression.Call(
+                    null,
+                    CachedReflectionInfo.TrySingle_TSource_1(typeof(TSource)), source.Expression));
+        }
+
```
Comment (JonHanna) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: `TrySingle` should probably work analogously to the other query methods; produce an expression representing a call of themselves on the expression of the source queryable. That way providers work on knowing what "TrySingle" means (or don't and fail, but the failure is on them). In particular the `EnumerableQuery` class will rewrite the call to `Queryable.TrySingle` as a call to `Enumerable.TrySingle` on the wrapped enumerable, and then work correctly.

Comment (JonHanna) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: (Also if you put a P2P reference into the queryable tests project to System.Linq you should be immediately able to test `TrySingle` on queryables produced by calling `AsQueryable()` on enumerables).

Comment (AlexRadch) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: TrySingle is different from other methods that it return boolean result and return out parameter. I can not create `Funct<T1, T2, TResult>` with out parameter. `Funct<T1, out T2, TResult>` is wrong for compiler.
Now I am waiting answer from @bartdesmet on this issue.

Comment (JonHanna) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: Ah. Yes, now that is a darn nuisance.

Comment (JonHanna) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: Wait. The `Func<>` is just a means to the end of getting a `MethodInfo`. Any matching delegate would offer us the same. Or am I missing something?  @bartdesmet?

Comment (AlexRadch) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: I will try decide this issue.

File: src/System.Linq/src/System/Linq/Single.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Single.cs]: Why not just have all of the logic in a private TrySingle that returns a tri-state rather than a bool?

Comment (AlexRadch) [inline src/System.Linq/src/System/Linq/Single.cs]: thanks for advice

File: src/System.Linq.Queryable/src/System/Linq/Queryable.cs (hunk 5 above)
Comment (bartdesmet) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: Another delegate would work here, but we'd still have a challenge with calling Execute<T> on the query provider given such an expression with an out parameter.

**Comment by karelz** [comment]
Marking as blocked to wait on the decision for method names TrySingle vs. TryGetSingle.

File: src/System.Linq.Queryable/src/System/Linq/Queryable.cs (hunk 5 above)
Comment (bartdesmet) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: One more note on my observation above.

The fact that `IQueryable<T>` implements `IEnumerable<T>` makes it equivalent to a `Func<IEnumerator<T>>` by means of having the `GetEnumerator` method. So it really *is* a `Compile` method in disguise, concretely implemented by the query provider author behind the `GetEnumerator` (and sometimes more deferred, behind the first `MoveNext` method, through an iterator) by passing the `Expression` to the `Execute` method.

In practice, the execution often involves the creation of a `Func<IEnumerable<T>>` through expression compilation etc., to include a result materialization step (e.g. the "final" `Select` clause causing object creation) and/or a partial local evaluation. So it really is using `Execute` as if there's a `Compile` step returning a compiled execution plan. This is often accompanied by other tricks to be able to reuse the same "plan" (by lifting out parameters that may change, e.g. from locally evaluated closures, and caching the resulting `Func<IEnumerable<T>>`) across multiple enumerations over the same query result.

This just shows that the aggregate methods really short-circuit a lot of phasing that's otherwise available through the laziness of `IEnumerable<T>`. For example:

```csharp
IQueryable<int> res = xs.Where(...).Select(...);
while (true)
{
int sum = res.Sum();
}
```
Short of playing caching tricks deep inside the provider, repeated evaluation of `Sum` does not give the query provider an opportunity to cache the "execution plan". It will be given a new expression tree each and every time. Contrast this to:

```csharp
IQueryable<int> res = xs.Where(...).Select(...);
while (true)
{
    foreach (int x in res) { ... }
}
```
In here, every repeated enumeration of result sequence can leverage the same plan which can be stored with the `IQueryable<T>` instance in `res`. This is a trick I play quite often in the stuff I'm working on here in Bing. The first `GetEnumerator` can prepare/compile the plan; subsequent ones can leverage it again. It does so by lifting out pieces that have to be re-evaluated locally each time (such as closed-over variables) through "funclet" expressions and generating a parameterized `Func<..., IEnumerator<T>>` factory which encapsulates the cached template that will be sent to the service over and over again (along with freshly evaluated arguments from the "funclets") upon repeated evaluation.

Alas, `Execute` does not provide an such option, because it doesn't return a `Quote<int>` like type (which would have an `Evaluate` method to get the `int`, and thus would be an object that can cache the "prepared" execution plan). Short of generalized quotes, even an `Execute` that'd return a `Func<T>` (or more generally, as described above, a `TDelegate`) would result in an object that can encapsulate cached execution plans for repeated evaluation (though it'd lack the ability to do further composition of the result of aggregates):

```csharp
Func<int> sum = res.Where(...).Select(...).Sum();
while (true)
{
    sum(); // no re-processing of a newly created expression tree over and over again
}
```

Also adding @divega who can chime in on "tales from the Entity Framework trenches" and offer an opinion on which way we should take `Queryable.TrySingle` so they can implement support for it.

File: src/System.Linq/src/System/Linq/ElementAt.cs [hunk 6]
```diff
@@ -10,51 +10,22 @@ public static partial class Enumerable
...
-                            }
-
-                            index--;
-                        }
-                    }
-                }
-            }
+            TSource result;
+            if (TryElementAt(source, index, out result))
+                return result;
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/ElementAt.cs]: Nit: please use Allman-style braces here.

```cs
if (TryElementAt(source, index, out result))
{
    return result;
}
```

File: src/System.Linq/src/System/Linq/ElementAt.cs [hunk 7]
```diff
@@ -64,7 +35,8 @@ public static TSource ElementAtOrDefault<TSource>(this IEnumerable<TSource> sour
             if (partition != null)
             {
                 bool found;
-                return partition.TryGetElementAt(index, out found);
+                element = partition.TryGetElementAt(index, out found);
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/ElementAt.cs]: @JonHanna After this PR is merged, would it make sense to change `IPartition` to follow a more conventional pattern of returning a bool and out-ing a `TSource`, and change all of its names from `TryGet*` to `Try*`?

File: src/System.Linq/src/System/Linq/First.cs [hunk 8]
```diff
@@ -10,43 +10,37 @@ public static partial class Enumerable
...
-            bool found;
-            TSource first = source.TryGetFirst(out found);
-            
-            if (!found)
-            {
-                throw Error.NoElements();
-            }
+            TSource result;
+            if (TryFirst(source, out result))
+                return result;
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/First.cs]: Nit: please use Allman-style braces here

File: src/System.Linq/src/System/Linq/First.cs [hunk 9]
```diff
@@ -10,43 +10,37 @@ public static partial class Enumerable
...
-            bool found;
-            TSource first = source.TryGetFirst(predicate, out found);
-
-            if (!found)
-            {
-                throw Error.NoMatch();
-            }
+            TSource result;
+            if (TryFirst(source, predicate, out result))
+                return result;
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/First.cs]: Nit: please use Allman-style braces here

File: src/System.Linq/src/System/Linq/Last.cs [hunk 10]
```diff
@@ -10,43 +10,37 @@ public static partial class Enumerable
...
-            bool found;
-            TSource last = source.TryGetLast(out found);
-            
-            if (!found)
-            {
-                throw Error.NoElements();
-            }
+            TSource result;
+            if (TryLast(source, out result))
+                return result;
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Last.cs]: Nit: please use Allman-style braces here

File: src/System.Linq/src/System/Linq/Last.cs [hunk 11]
```diff
@@ -10,43 +10,37 @@ public static partial class Enumerable
...
         {
-            bool found;
-            TSource last = source.TryGetLast(predicate, out found);
-
-            if (!found)
-            {
-                throw Error.NoMatch();
-            }
+            TSource result;
+            if (TryLast(source, predicate, out result))
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Last.cs]: Nit: please use Allman-style braces here

File: src/System.Linq/src/System/Linq/Last.cs [hunk 12]
```diff
@@ -107,49 +103,45 @@ internal static TSource TryGetLast<TSource>(this IEnumerable<TSource> source, Fu
...
             {
-                return ordered.TryGetLast(predicate, out found);
+                bool found;
+                element = ordered.TryGetLast(predicate, out found);
+                return found;
             }
-
-            IList<TSource> list = source as IList<TSource>;
-            if (list != null)
+            else
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Last.cs]: Why did you add an unnecessary `else` here and indent everything?

File: src/System.Linq/src/System/Linq/Last.cs [hunk 13]
```diff
@@ -107,49 +103,45 @@ internal static TSource TryGetLast<TSource>(this IEnumerable<TSource> source, Fu
...
+                            element = item;
+                            return true;
+                        }
                     }
                 }
-            }
-            else
-            {
-                using (IEnumerator<TSource> e = source.GetEnumerator())
+                else
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Last.cs]: Same here: Why do you need to indent everything?

File: src/System.Linq/src/System/Linq/Last.cs [hunk 14]
```diff
@@ -107,49 +103,45 @@ internal static TSource TryGetLast<TSource>(this IEnumerable<TSource> source, Fu
...
                 {
-                    while (e.MoveNext())
+                    TSource result = default(TSource);
+                    bool found = false;
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Last.cs]: Nit: unneeded local variables. You can replace the `result = item; found = true` with `element = item; return true` and delete both of these variables.

File: src/System.Linq/src/System/Linq/ElementAt.cs (hunk 7 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/ElementAt.cs]: I should say so. The rather unusual way round they are is an artefact of the incremental way they were arrived at rather than for any other reason (I tried changing before and it had a tiny negative impact on performance, but I wouldn't worry about it if we'd a better reason for doing the change as we do now, and also it's likely going to be a small improvement now that it matches more code at the public interface).

File: src/System.Linq/src/System/Linq/Last.cs (hunk 14 above)
Comment (karelz) [inline src/System.Linq/src/System/Linq/Last.cs]: Advise: Don't think about performance, measure it first instead - always, no exceptions.
Also avoid premature optimizations (optimizing something that does not matter - e.g. saving single instruction may not be the right trade-off for code complexity, additional allocation, or more JITTing.

In this particular case, you have to pick the level of abstraction - you have to rely on JIT doing the right thing. If you have evidence it doesn't, file a bug against it, but do not micro-optimize code to compensate for that, unless it is the most perf-sensitive thing in your project.

Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Last.cs]: @AlexRadch Ah, I see what you mean; if we assign to the out parameter and `predicate` or the enumerator throws an exception, we will have assigned a overwritten the value at `element`. In that case, 👍. Might want to consider adding a test for that.

File: src/System.Linq/src/System/Linq/Single.cs [hunk 15]
```diff
@@ -91,8 +93,16 @@ public static TSource SingleOrDefault<TSource>(this IEnumerable<TSource> source)
...
+                    case 0:
+                        {
+                            element = default(TSource);
+                            return TrySingleResult.NoElements;
+                        }
+                    case 1:
+                        {
+                            element = list[0];
+                            return TrySingleResult.SingleElement;
+                        }
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Single.cs]: Nit: Do not indent/add brackets in case clauses

```diff
+                    case 1:
+                        element = list[0];
+                        return TrySingleResult.SingleElement;
```

File: src/System.Linq/src/System/Linq/Last.cs [hunk 16]
```diff
@@ -107,49 +107,43 @@ internal static TSource TryGetLast<TSource>(this IEnumerable<TSource> source, Fu
...
+                        return true;
                     }
                 }
             }
             else
             {
-                using (IEnumerator<TSource> e = source.GetEnumerator())
+                TSource result = default(TSource);
+                bool found = false;
+                foreach (TSource item in source)
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Last.cs]: Nit: Can you add a newline between the variable declarations and `foreach`?

File: src/System.Linq.Queryable/src/System/Linq/Queryable.cs (hunk 5 above)
Comment (AlexRadch) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: @bartdesmet If you did not find solution than I can suggest next. Make methods with Tuple in Enumerable public but not visible to editor. See EditorBrowsable(EditorBrowsableState.Never) attribute

Comment (divega) [inline src/System.Linq.Queryable/src/System/Linq/Queryable.cs]: @bartdesmet thanks for binging this to my attention. I have provided my general feedback on the feature at https://github.com/dotnet/corefx/issues/6073#issuecomment-266867512. 

Besides that, we seem to share many of the same experience trying to overcome the limitations of the `IQueryable`/`IQueryProvider` story, e.g. how the ability to compose expressions for deferred execution is conflated with types representing collections. But this matter seems to go way beyond what this PR is about so I believe it would be great if we could have a separate discussion on it one of these days.

**Comment by VSadov** [comment]
@bartdesmet - we should have a separate discussion about aggregating in a quoted space. I agree the forced evaluation may be inconvenient.

**Comment by VSadov** [comment]
There is clearly a problem with parallel implementation of TrySingle in both IEnumerable and IQueryable

Normally we would have isomorphic operators on both sides and Queryable factory would just refer to itself in the expression nodes, which can be trivially handled by the EnumerableQuery/Rewriter. 
Due to this duality the current shape of the API would call for the use of "out" parameter on both sides.

The problem is that existing "IQueryProvider.Execute" returns only a single value - sequence or an element. Introducing additional API just for this change and in particular with an "out" parameter does not seem very appealing. It is not even clear whether it would be composable enough.

While in theory a sequence could be used here, with an empty sequence indicating a failure, it would seem that such a solution is not all that far from the original problem. (it might still be an additional idea to think about...)

Considering that the only purpose of the "out" parameter here is to return 2 values - Result and the "success" flag, and in the light of the above problems, perhaps changing the API to return a tuple seems more appropriate. After all returning multiple values is exactly the scenario where tuples are supposed to help.

So, what do you think about changing the API (for both Enumerable and Queryable) to return a `ValueTuple<T, bool>` instead? 

Or more precisely a tuple ` (T Result, bool Succeded)`   , with actual element order and names TBD.

**Comment by jamesqo** [comment]
> So, what do you think about changing the API (for both Enumerable and Queryable) to return a `ValueTuple<T, bool>` instead?

@VSadov There was a discussion on that in the linked proposal-- see https://github.com/dotnet/corefx/issues/6073#issuecomment-266867512. I suggested we could add an `Optional<T>` type to `System.Linq`. It would be like `Nullable`, except without the artificial struct constraint, and you could do things like `Select` / `SelectMany` on it since it would implement the Maybe monad. `ValueTuple<T, bool>` IMO would be a little awkward even with named tuples.

**Comment by bartdesmet** [comment]
I'm personally in favor of value tuples because they're easy to consume going forward and will work well with existing queryable infrastructure.

If an Optional<T> type is considered, it'd be a pity if it is buried in the System.Linq namespace given its general use (e.g. Roslyn has similar types floating around). This said, it'd look quite asymmetric compared to its Nullable<T> analog due to missing language convenience such as equality checks with null, conditional access operators (a maybe monad in disguise for null propagation), lifting, etc. These may not be deal breakers but I feel a separate discussion on the merits of Optional<T> in a more appropriate place in the BCL (e.g. System) would be in order, taking the Nullable<T> analogy in consideration, etc.

**Comment by jamesqo** [comment]
@bartdesmet I have thought about this some more. I am changing my mind-- `ValueTuple` seems to be OK to use here, especially with the language integration. I think introducing an `Optional<T>` type would probably confuse people about using it vs. Nullable, and it would have a very limited/avoidable use case (just to avoid the `Try*(out T)` pattern in generic code), this being one of the exceptions. There is already so much compiler integration being done around nullables/tuples (e.g. null-conditional) that this would only serve to go against the grain. So +1 for tuples.

**Comment by MadsTorgersen** [comment]
I commented over in https://github.com/dotnet/corefx/issues/6073#issuecomment-270228403, essentially to the effect that we should have a tuple-based overload for encoding in expression trees, and a preferred out-based overload for users to call.

As for _which_ tuple return mechanism is used, it matters less if that overload is primarily for the benefit of query providers, not callers. I don't have a strong opinion, but:
- In this setup, a new `Option<T>` type or equivalent is certainly not warranted
- The usage benefits of ValueTuple<T>'s language support over Tuple<T> matter little, but
- the performance benefits may matter

**Comment by karelz** [comment]
I assume this PR is abandoned as other @AlexRadch's PRs. Closing for now.
If it is not the case, please let me know. I'll be happy to reopen it.
